### ISSUE TITLE
Cilium still sends traffic to a Terminating pod (managed by a StatefulSet) on a lost node 

### ISSUE THREAD

**Issue by alexandrevilain** [cilium/cilium#21763]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

We found an issue when using a service targeting a statefulset when a node is lost. 
When one of the statefulset member is on a lost node, cilium continues to load balance traffic to the Terminating pod so the request hangs. 
It looks like the CilliumEndpoint stays in the `ready` state.

To reproduce: 
```
kubectl apply -f https://gist.githubusercontent.com/alexandrevilain/a991425ca7439318a96a00a4392b118a/raw/0443ad8acb6671c9e878ee9f427968fad369803e/cilliumissue.yaml
```

It creates a statefulset with 3 running pods:

```
$ kubectl get pods -n ciliumissue -o wide
NAME    READY   STATUS    RESTARTS   AGE     IP              NODE                   NOMINATED NODE   READINESS GATES
web-0   1/1     Running   0          7m48s   192.168.2.51    cluster-02-worker-02   <none>           <none>
web-1   1/1     Running   0          7m47s   192.168.3.119   cluster-02-worker-01   <none>           <none>
web-2   1/1     Running   0          7m44s   192.168.1.6     cluster-02-worker-03   <none>           <none>
```

Here using a pod like `netshoot`, running `curl http://nginx` against the `nginx` service works like a charm. 
Then, break the node. On our side, we break the rke2 start command and do an hard reboot on openstack.

The node becomes `NotReady`:

```
$ kubctl get nodes
NAME                   STATUS     ROLES                       AGE   VERSION
cluster-02-master-01   Ready      control-plane,etcd,master   12d   v1.24.4+rke2r1
cluster-02-master-02   Ready      control-plane,etcd,master   12d   v1.24.4+rke2r1
cluster-02-master-03   Ready      control-plane,etcd,master   12d   v1.24.4+rke2r1
cluster-02-worker-01   Ready      <none>                      12d   v1.24.4+rke2r1
cluster-02-worker-02   NotReady   <none>                      12d   v1.24.4+rke2r1
cluster-02-worker-03   Ready      <none>                      12d   v1.24.4+rke2r1
```

Then wait for the pod to be reported in `Terminating` state:

```
$ kubectl get pods -n ciliumissue -o wide
NAME    READY   STATUS        RESTARTS   AGE     IP              NODE                   NOMINATED NODE   READINESS GATES
web-0   1/1     Terminating   0          7m52s   192.168.2.51    cluster-02-worker-02   <none>           <none>
web-1   1/1     Running       0          7m51s   192.168.3.119   cluster-02-worker-01   <none>           <none>
web-2   1/1     Running       0          7m48s   192.168.1.6     cluster-02-worker-03   <none>           <none>
```

Then start a new `netshoot` pods. The previous `curl` command will hangs randomly. 

If we look at the `CiliumEndpoints`, it's still in ready state:

```
$ kubectl get ciliumendpoints.cilium.io -n ciliumissue
NAME    ENDPOINT ID   IDENTITY ID   INGRESS ENFORCEMENT   EGRESS ENFORCEMENT   VISIBILITY POLICY   ENDPOINT STATE   IPV4            IPV6
debug   2983          55204                                                                        ready            192.168.3.46
web-0   1935          16310                                                                        ready            192.168.2.51
web-1   292           7027                                                                         ready            192.168.3.119
web-2   2852          8650                                                                         ready            192.168.1.6
```

I reproduced this issue many times using the provided manifest. 

### Cilium Version

Client: 1.12.0 9447cd1 2022-07-19T12:22:00+02:00 go version go1.18.4 linux/amd64
Daemon: 1.12.0 9447cd1 2022-07-19T12:22:00+02:00 go version go1.18.4 linux/amd64

### Kernel Version

5.15.0-48-generic

### Kubernetes Version

v1.24.4+rke2r1

### Anything else?

Maybe it can be related to: https://github.com/cilium/cilium/issues/19931 but it's not the exact same behavior. 

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by aditighag** [comment]
👋 Does the pod in question have liveness probes configured? If yes, what's the probes frequency configured? Also, how long did you keep on checking to see that the traffic was being load-balanced to the endpoints on lost nodes?

To give some more context, Cilium relies on k8s to set endpoints state to terminating before it can gracefully terminate such endpoints for load-balancing [1]. If cilium pods running on other nodes don't receive these k8s watcher events for the endpoints deployed on the lost node, then they will be unaware of the terminating endpoints. This probably explains why you also see those `CiliumEndpoints` in ready state. Also, CiliumEndpoint state isn't considered for load-balancing. That's a Cilium specific custom resource name.

Can you check if cilium pods on other nodes are getting the terminating endpoints events from Kubernetes? Based on the documentation [1], you can also check for the metrics - `k8s_terminating_endpoints_events_total` as a first step. 

[1] https://docs.cilium.io/en/v1.12/gettingstarted/kubeproxy-free/#graceful-termination

**Comment by alexandrevilain** [comment]
Hi @aditighag ! Thanks for your answer 👍 

First of all, I found the issue on a pod with the following liveness probe configured:
```yaml
livenessProbe:
  failureThreshold: 8
  httpGet:
    path: /health
    port: 2381
    scheme: HTTP
  initialDelaySeconds: 10
  periodSeconds: 10
  successThreshold: 1
  timeoutSeconds: 15
```

I don't have liveness probe on my manifests to reproduce it. But I'll try with a liveness probe right now.

It seems that other nodes are getting the informations for terminating endpoints:

```
$ curl --silent http://10.0.40.3:9962/metrics | grep k8s_terminating_endpoints_events_total
# HELP cilium_k8s_terminating_endpoints_events_total Number of terminating endpoint events received from Kubernetes
# TYPE cilium_k8s_terminating_endpoints_events_total counter
cilium_k8s_terminating_endpoints_events_total 319
```

I'm now trying my example manifest with liveness probe, I'll get back to you ASAP ;)

**Comment by aanm** [comment]
@alexandrevilain can you provide a sysdump of the cluster where the problem is happening? Thank you

**Comment by alexandrevilain** [comment]
Hi! Sorry for the delay.

I reproduced the issue on a brand new cluster using kubeadm (kubernetes v1.24.4) with cilium v1.12.4 (`kubeProxyReplacement: strict`) running on Openstack.

I deployed the previously provided nginx statefulset and broke the `cluster-01-worker-01`.

When the nodes becomes `NotReady` running `curl nginx.default` in a debug pod works well. But when the Pod running on the broken node goes to `Terminating` state, the traffic starts to hangs for some requests.

Here is the sysdump: 

[cilium-sysdump-20221125-104520.zip](https://github.com/cilium/cilium/files/10091757/cilium-sysdump-20221125-104520.zip)
